libvoikko/src/voikkospell.1
.\"                                      Hey, EMACS: -*- nroff -*-
.\" First parameter, NAME, should be all caps
.\" Second parameter, SECTION, should be 1-8, maybe w/ subsection
.\" other parameters are allowed: see man(7), man(1)
.TH VOIKKOSPELL 1 "2008-07-01"
.\" Please adjust this date whenever revising the manpage.
.\"
.\" Some roff macros, for reference:
.\" .nh        disable hyphenation
.\" .hy        enable hyphenation
.\" .ad l      left justify
.\" .ad b      justify to both left and right margins
.\" .nf        disable filling
.\" .fi        enable filling
.\" .br        insert line break
.\" .sp <n>    insert n+1 empty lines
.\" for manpage-specific macros, see man(7)
.SH NAME
voikkospell \- test program for Voikko spell checker
.SH SYNOPSIS
.B voikkospell
.RI [ options ]
.SH DESCRIPTION
.B voikkospell
is a test program for spell checking functionality in libvoikko,
library of Finnish language tools.
It reads words from stdin (one word on a line) and print them to stdout, adding
"C: " in front of correct words and "W: " in front of incorrect words. voikkohyphenate prints
words in fully hyphenated form.

Common options of different Voikko test programs are listed in COMMON OPTIONS.
.SH OPTIONS
.TP
.B \-t
Prints only "C" or "W" instead of typical full output.
.TP
.B \-s
Prints suggestions (S:) for incorrectly spelled words.
.TP
.BI \-c n
Set cache size parameter to
.IR n .
\-1 disables the cache and 0 is the default.
For checking large amounts of unsorted text you may want to set
.I n
to 5 to get better performance.
.TP
.BI \-x c
Like
.B voikkospell -s
but output is printed on one line separated by character
.IR c 
without "C", "W" or "S" in front of the words.
If
.IR c
is not defined words are separated by space and suggestions that have
spaces in them are not printed.
.TP
.BI ignore_nonwords= n
.TP
.BI accept_first_uppercase= n
.TP
.BI accept_extra_hyphens= n
.TP
.BI accept_missing_hyphens= n
.TP
.BI ocr_suggestions= n
.br
Set the value of the specified boolean option.
.I n
can be either 0 (false) or 1 (true).
.SH COMMON OPTIONS
.TP
.BI \-p " directory"
(voikkospell, voikkohyphenate)
Look from
.I directory
before the standard locations when searching for dictionary files.
.TP
.BI ignore_dot= n
(voikkospell, voikkohyphenate)
.TP
.BI ignore_numbers= n
(voikkospell, voikkohyphenate)
.br
Set the value of the specified boolean option.
.I n
can be either 0 (false) or 1 (true).
.SH AUTHOR
voikkospell and this manual page were written by Harri Pitk\[:a]nen (hatapitk@iki.fi).


libvoikko/ChangeLog
2008-07-03 Harri Pitkänen <hatapitk@iki.fi>

	* Work on grammar checker.
	* Add whitespace error checks.

2008-07-02 Harri Pitkänen <hatapitk@iki.fi>

	* Work on grammar checker.
	* Increase library version number to 2.0.
	* Update README.

2008-07-01 Harri Pitkänen <hatapitk@iki.fi>

	* Split and update manual pages.
	* Add single paragraph grammar checker cache.

2008-06-29 Harri Pitkänen <hatapitk@iki.fi>

	* Move grammar checking related functionality from
	  voikkohyphenate to a new test tool voikkogc.

2008-06-28 Harri Pitkänen <hatapitk@iki.fi>

	* Add API for grammar checking.

2008-06-08 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.
	* Allow strings longer than LIBVOIKKO_MAX_WORD_CHARS to
	  be converted with voikko_cstrtoucs4.

2008-06-07 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.

2008-06-04 Harri Pitkänen <hatapitk@iki.fi>

	* Work on sentence splitting.
	* Increase version numbers for version 1.8.

2008-05-26 Ville-Pekka Vainio <vpivaini@cs.helsinki.fi>

	* Provide libvoikko.pc for pkg-config.

2008-03-07 Harri Pitkänen <hatapitk@iki.fi>

	* Fix hyphenation of nougat'kaan.

2008-01-29 Harri Pitkänen <hatapitk@iki.fi>

	* Increase cost limit for suggestions from 250 to 350.

2008-01-19 Harri Pitkänen <hatapitk@iki.fi>

	* Iconv on OS X should use UCS-4-INTERNAL too.

2007-12-30 Harri Pitkänen <hatapitk@iki.fi>

	* Some fixes for VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS.
	* Add option VOIKKO_OPT_ACCEPT_MISSING_HYPHENS.

2007-12-29 Harri Pitkänen <hatapitk@iki.fi>

	* Add option VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS.

2007-12-01 Harri Pitkänen <hatapitk@iki.fi>

	* Tokenization functions should not be used without initializing the
	  library.

2007-11-28 Harri Pitkänen <hatapitk@iki.fi>

	* Fix configure test AM_ICONV failing with -Wall on FreeBSD.
	* Tweak URL recognition so that "J.K.Paasikivi" will not be
	  considered as hostname.

2007-11-27 Harri Pitkänen <hatapitk@iki.fi>

	* Add boolean option VOIKKO_OPT_IGNORE_NONWORDS.
	* Do not link voikkospell and voikkohyphenate against libmalaga.

2007-10-30 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-28 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-27 Harri Pitkänen <hatapitk@iki.fi>

	* Additional work on string tokenizer.

2007-10-21 Harri Pitkänen <hatapitk@iki.fi>

	* Allow the use of environment variable VOIKKO_DICTIONARY_PATH
	  to specify the dictionary path.

2007-10-15 Harri Pitkänen <hatapitk@iki.fi>

	* Add preliminary version of string tokenizer
	  voikko_next_token_ucs4.
	* Increase version numbers for 1.6.

2007-09-08 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement suggestion "." -> ","

2007-08-23 Harri Pitkänen <hatapitk@iki.fi>

	* Change configure.ac to fix build on FreeBSD

2007-08-20 Harri Pitkänen <hatapitk@iki.fi>

	* FreeBSD fix, thanks to Jaakko Heinonen

2007-07-08 Harri Pitkänen <hatapitk@iki.fi>

	* Add option ignore_numbers=n to voikkospell. Change
	  default from 1 to 0 (the library default).
	* Undo 1-to-1 normalisations in spelling suggestions.

2007-06-19 Harri Pitkänen <hatapitk@iki.fi>

	* Add exported functions for freeing the memory
	  allocated for result structures. This is needed
	  in environments where multiple C runtime libraries
	  are used in one process.
	* Increase version numbers for 1.5

2007-06-17 Harri Pitkänen <hatapitk@iki.fi>

	* Generate libvoikko-1.def on Windows to enable
	  dynamic linking of libvoikko-1.dll with code
	  generated using MSVC.
	* Set Finnish locale on Windows to ensure that
	  character case conversions work correctly.

2007-05-07 Harri Pitkänen <hatapitk@iki.fi>

	* Hyphenating "vast'edes" as "vast'e-des" is ugly
	* Hyphenating after ' is wrong
	* Increase version numbers for 1.4.1

2007-04-19 Harri Pitkänen <hatapitk@iki.fi>

	* Check for embedded dots in URL detection
	* Only use URL and number detection when
	  VOIKKO_OPT_NO_UGLY_HYPHENATION is set

2007-04-07 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement _ -> (any character) to OCR suggestions 

2007-03-28 Harri Pitkänen <hatapitk@iki.fi>

	* Update to Automake 1.10
	* Update version information for version 1.4
	* Avoid using rule based hyphenator with certain strings containing
	  numbers and special characters

2007-02-26 Harri Pitkänen <hatapitk@iki.fi>

	* Fix the option VOIKKO_INTERSECT_COMPOUND_LEVEL, which only worked
	  correctly with values 0 and 1. Bug report and suggested fix by
	  Kalle Lampila.
	* Add option intersect_compound_level=n to voikkohyphenate

2007-02-16 Harri Pitkänen <hatapitk@iki.fi>

	* Add replacement b <-> h to OCR suggestions

2007-02-14 Harri Pitkänen <hatapitk@iki.fi>

	* New command line option -x for voikkospell, from a slightly
	  modified patch by Hannu Väisänen
	* Try replacing 0 with o when generating OCR suggestions

2007-02-09 Harri Pitkänen <hatapitk@iki.fi>

	* Fix broken suggestions for words with non-default capitalisation
	* No longer leak iconv handles when library initialisation fails
	* Increase version to 1.3.1

2007-01-28 Harri Pitkänen <hatapitk@iki.fi>

	* Fix character set conversion on Windows

2007-01-27 Harri Pitkänen <hatapitk@iki.fi>

	* Fix typo in voikko_hyphenate.h

2007-01-26 Harri Pitkänen <hatapitk@iki.fi>

	* Add new option VOIKKO_MIN_HYPHENATED_WORD_LENGTH.
	* Remove empty TODO.

2007-01-25 Harri Pitkänen <hatapitk@iki.fi>

	* Add some more hyphenation points when VOIKKO_OPT_NO_UGLY_HYPHENATION
	  is not set.

2007-01-20 Harri Pitkänen <hatapitk@iki.fi>

	* Add a word length check to voikko_split_compounds.
	* Fixes to documentation.

2007-01-16 Harri Pitkänen <hatapitk@iki.fi>

	* Support explicitly specified string lengths as opposed to only
	  allowing null terminated strings in certain utility functions.
	* VOIKKO_OPT_IGNORE_DOT is now respected during hyphenation as well,
	  and corresponding option is supported by voikkohyphenate. This
	  was needed because apparently at least OpenOffice.org feeds the

libvoikko/src/voikkospell.c
/* Voikkospell: Testing tool for libvoikko
 * Copyright (C) 2006 Harri Pitkänen <hatapitk@iki.fi>
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the GNU General Public License
 * as published by the Free Software Foundation; either version 2
 * of the License, or (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
 *********************************************************************************/

#define _GNU_SOURCE

#include <voikko.h>
#include <stdio.h>
#include <stdlib.h>
#include <unistd.h>
#include <locale.h>
#include "porting.h"
#ifdef HAVE_NL_LANGINFO
#include <langinfo.h>
#endif // HAVE_NL_LANGINFO
#include <stdio.h>
#include <string.h>

int autotest = 0;
int suggest = 0;
int one_line_output = 0;
char word_separator = ' ';
int space = 0;  /* Set to nonzero if you want to output suggestions that has spaces in them. */

void check_word(int handle, char * word) {
	char ** suggestions;
	int i;
	int result = voikko_spell_cstr(handle, word);
	if (result == VOIKKO_CHARSET_CONVERSION_FAILED) {
		printf("E: charset conversion failed\n");
		return;
	}
	if (result == VOIKKO_INTERNAL_ERROR) {
		printf("E: internal error\n");
		return;
	}
	if (autotest) {
		if (result) printf("C\n");
		else printf("W\n");
		fflush(0);
	}
	else if (one_line_output) {
		printf ("%s", word);
		if (!result) {
			suggestions = voikko_suggest_cstr(handle, word);
			if (suggestions) {
				for (i = 0; suggestions[i] != 0; i++) {
					if (space || (!space && strchr (suggestions[i], ' ') == 0))
						printf ("%c%s", word_separator, suggestions[i]);
				}
				voikko_free_suggest_cstr(suggestions);
			}
		}
		printf ("\n");
	}
	else {
		if (result) printf("C: %s\n", word);
		else printf("W: %s\n", word);
	}
	if (!one_line_output && suggest && !result) {
		suggestions = voikko_suggest_cstr(handle, word);
		if (suggestions) {
			for (i = 0; suggestions[i] != 0; i++) {
				printf("S: %s\n", suggestions[i]);
			}
			voikko_free_suggest_cstr(suggestions);
		}
	}
}



int main(int argc, char ** argv) {
	size_t size = LIBVOIKKO_MAX_WORD_CHARS;
	char * line;
	ssize_t chars_read;
	char * encoding;
	char * path = 0;
	int handle;
	int i;
	int cache_size;

	line = malloc(size);
	if (line == 0) {
		printf("E: Out of memory\n");
		return 1;
	}
	
	cache_size = 0;
	for (i = 1; i < argc; i++) {
		if (strncmp(argv[i], "-c", 2) == 0) {
			cache_size = atoi(argv[i] + 2);
		}
		else if (strcmp(argv[i], "-p") == 0 && i + 1 < argc) path = argv[++i];
	}
	const char * voikko_error = (const char *) voikko_init_with_path(&handle, "fi_FI", cache_size, path);
	if (voikko_error) {
		printf("E: Initialisation of Voikko failed: %s\n", voikko_error);
		free(line);
		return 1;
	}
	
	setlocale(LC_ALL, "");
	encoding = nl_langinfo(CODESET);
	
	voikko_set_string_option(handle, VOIKKO_OPT_ENCODING, encoding);
	
	for (i = 1; i < argc; i++) {
		if (strcmp(argv[i], "-t") == 0) autotest = 1;
		else if (strcmp(argv[i], "ignore_dot=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_DOT, 1);
		else if (strcmp(argv[i], "ignore_dot=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_DOT, 0);
		else if (strcmp(argv[i], "ignore_numbers=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_NUMBERS, 1);
		else if (strcmp(argv[i], "ignore_numbers=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_NUMBERS, 0);
		else if (strcmp(argv[i], "ignore_nonwords=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_NONWORDS, 1);
		else if (strcmp(argv[i], "ignore_nonwords=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_IGNORE_NONWORDS, 0);
		else if (strcmp(argv[i], "accept_first_uppercase=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_FIRST_UPPERCASE, 1);
		else if (strcmp(argv[i], "accept_first_uppercase=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_FIRST_UPPERCASE, 0);
		else if (strcmp(argv[i], "accept_extra_hyphens=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS, 1);
		else if (strcmp(argv[i], "accept_extra_hyphens=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_EXTRA_HYPHENS, 0);
		else if (strcmp(argv[i], "accept_missing_hyphens=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_MISSING_HYPHENS, 1);
		else if (strcmp(argv[i], "accept_missing_hyphens=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_ACCEPT_MISSING_HYPHENS, 0);
		else if (strcmp(argv[i], "ocr_suggestions=1") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_OCR_SUGGESTIONS, 1);
		else if (strcmp(argv[i], "ocr_suggestions=0") == 0)
			voikko_set_bool_option(handle, VOIKKO_OPT_OCR_SUGGESTIONS, 0);
		else if (strncmp(argv[i], "-x", 2) == 0) {
			one_line_output = 1;
			if (strlen(argv[i]) == 3) word_separator = argv[i][2];
			space = (word_separator != ' ');
		}
		else if (strcmp(argv[i], "-s") == 0) suggest = 1;
	}
	
	while (1) {
		chars_read = getline(&line, &size, stdin);
		if (chars_read == -1) break;
		if (chars_read > 0 && line[chars_read - 1] == '\n') {
			line[chars_read - 1] = '\0';
		}
		check_word(handle, line);
	}
	free(line);
	voikko_terminate(handle);
	return 0;
}

